src/librustc_mir/interpret/memory.rs

```diff
@@ -53,6 +54,85 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     pub tcx: TyCtxtAt<'a, 'tcx, 'tcx>,
 }
 
+impl<'a, 'mir, 'tcx, M> Clone for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{
+    fn clone(&self) -> Self {
+        Memory {
+            data: self.data.clone(),
+            alloc_kind: self.alloc_kind.clone(),
+            alloc_map: self.alloc_map.clone(),
+            uninitialized_statics: self.uninitialized_statics.clone(),
+            cur_frame: self.cur_frame.clone(),
+            tcx: self.tcx.clone(),
+        }
+    }
+}
+
+impl<'a, 'mir, 'tcx, M> Eq for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{}
+
+impl<'a, 'mir, 'tcx, M> PartialEq for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{
+    fn eq(&self, other: &Self) -> bool {
+        let Memory {
+            data,
+            alloc_kind,
+            alloc_map,
+            uninitialized_statics,
+            cur_frame,
+            tcx,
+        } = self;
+
+        *data == other.data
+            && *alloc_kind == other.alloc_kind
+            && *alloc_map == other.alloc_map
+            && *uninitialized_statics == other.uninitialized_statics
+            && *cur_frame == other.cur_frame
+            && ptr::eq(tcx, &other.tcx)
+    }
+}
+
+impl<'a, 'mir, 'tcx, M> Hash for Memory<'a, 'mir, 'tcx, M>
+    where M: Machine<'mir, 'tcx>,
+          'tcx: 'a + 'mir,
+{
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let Memory {
+            data,
+            alloc_kind: _,
+            alloc_map: _,
+            uninitialized_statics: _,
+            cur_frame,
+            tcx,
+        } = self;
+
+        data.hash(state);
+        cur_frame.hash(state);
+        (tcx as *const _ as usize).hash(state);
+
+        // We ignore some fields which don't change between evaluation steps.
+
+        // Since HashMaps which contain the same items may have different
+        // iteration orders, we use a commutative operation (in this case
+        // addition, but XOR would also work), to combine the hash of each
+        // `Allocation`.
+        self.allocations()
+            .map(|allocs| {
+                let mut h = FxHasher::default();
+                allocs.hash(&mut h);
```

Comment:
So this is problematic. Hashing an `AllocId` (inside a relocation of the alloc) will hash the actual `Id`. This means that if we ever call a function in a loop, since we keep getting new `AllocId`s for the function's locals, we'll never realize we're in a loop (ok we will outside the function, but imagine the function returning a raw pointer to its local, and we keep updating a field in the outer function).

I think it's fine for now, and will probably go away once we get around to implementing local `AllocId`s.

Reply:
So right now, this hashes *both* the `AllocId` and the `Allocation` (which can be crazy slow btw). Would simply hashing `Allocation` be correct?

I could maybe improve performance by memoizing the hash of an `Allocation`. This would involve setting a `Cell<Option<u64>>` to `None` whenever an `Allocation` is mutated, even when the loop detector isn't running. This might not be too bad, since there's no immediate data dependency on the cached value, and it will often be `None` during regular execution so the branch predictor will be on our side.

Reply:
We could benchmark it. For now let's do the slow version. We are already in "this might take a while territory"

Reply:
We might not even have to hash the memory, if we transitively hash through AllocIds by hashing their allocation instead of the id

Reply:
Agreed, I'm a sucker for a premature optimization :).

> transitively hash through AllocIds by hashing their allocation instead of the id

Can you elaborate on this? I don't quite understand.

Reply:
So. The frame locals can contain AllocIds. Instead of just hashing these normally, you'd have to hand roll recursive hashing code that, whenever it sees an AllocId, instead of hashing that u64, it loads the allocation from the memory and the hashes that the same way: any encountered AllocIds are instead used to hash their contents.

In order for that not to explode you need to keep around a `HashMap<AllocId, u64>` where the u64 are already hashed Allocation hashes, so you can just hash that hash instead of the entire allocation

Reply:
Okay, so the recursive part is the (`Relocations`)[https://github.com/rust-lang/rust/blob/441290291fc71dbfa16a372636a59da75649eece/src/librustc/mir/interpret/mod.rs#L480] field of `Allocation`.

Reply:
I'm caught up now; I'll repeat what you said for my own clarity. 

`alloc_map`, contains all `Allocation`s which are visible to a given `EvalContext` indexed by `AllocId`. The naive approach, comparing the `alloc_map`s with `eq`, will correctly identify some infinite loops. However,  if an object is reallocated, and all pointers to it are updated to point to the new value, the program states are equivalent but our equality function will fail.

Instead, we can look at the program heap as a graph of allocations whose roots are `Value::ByRef`s on the stack and whose edges are the `AllocId`s in `Relocations`. Equality over `EvalSnaphost` `Frame`s is the same except when it comes to these `ByRef` variants. We would push them onto a separate list for each frame, then do a simultaneous BFS/DFS, comparing each `Allocation` along the way. This would make equality invariant over changing `AllocId`s.

It is much easier to implement `Hash` than `PartialEq` with this approach: we simply omit any `AllocId`s--either on the stack in a `Value::ByRef` or the heap in a `Relocations`--from the hash.

I think I will implement this naively at first (not handling object reallocation), then try to come up with a test case where the more robust case would be necessary to detect infinite loops.

Reply:
Also, I think the result of casting a reference to `usize` during CTFE is still up in the air. If the `AllocId` is used in any way when computing that, the more complex strategy will no longer be correct.

Reply:
The following test case fails as you predicted.

```rust
#![feature(const_fn, const_let)]

const fn churn() -> *const u8 {
    let x = 0u8;
    &x
}

fn main() {
    let _ = [(); {
        let mut y = 0 as *const u8;
        loop {
            y = churn();
        }
        y as usize
    }];
}
```

There must be a subtler way of triggering this, but I can't come up with one at the moment. I don't think this class of programs is worth implementing a custom equality function which ignores `AllocId`s. Let me know if you disagree.

Reply:
We can at least check for each AllocId that we hash, whether it is dangling, and then just hash some default value (e.g. `u64::max_value`), unfortunately that requires the same changes as walking down AllocIds...

If you open an issue about the not working example and the explanation of the full fix, we can merge this PR as it is.

Reply:
I'll open an issue. I plan to keep working on this btw.